### PR TITLE
#56040 site health persistent object cache

### PR DESCRIPTION
**Pull request by manuelRod**
Adding Persistent Object cache check to Site Health.

Trac ticket: [#56040](https://core.trac.wordpress.org/ticket/56040)

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-admin/includes/class-wp-site-health.php [hunk 1]
```diff
@@ -2253,6 +2253,104 @@ public function get_test_authorization_header() {
...
+	/**
+	 * Tests persistent_object_cache health check.
+	 *
+	 * @since 6.1
+	 *
+	 * @return array The health check result suggesting persistent object caching.
+	 */
+	public function get_test_health_persistent_object_cache() {
+		/**
+		 * Filter the action URL for the persistent object cache health check.
```
Comment (audrasjb) [inline src/wp-admin/includes/class-wp-site-health.php]: `Filters` (as per coding standards, third person singular verbs should be used in function descriptions)

File: src/wp-admin/includes/class-wp-site-health.php [hunk 2]
```diff
@@ -2253,6 +2253,104 @@ public function get_test_authorization_header() {
 		return $result;
 	}
 
+	/**
+	 * Tests persistent_object_cache health check.
+	 *
+	 * @since 6.1
```
Comment (audrasjb) [inline src/wp-admin/includes/class-wp-site-health.php]: `@since 6.1.0` would be better

File: src/wp-admin/includes/class-wp-site-health.php [hunk 3]
```diff
@@ -2253,6 +2253,104 @@ public function get_test_authorization_header() {
...
+	 *
+	 * @return array The health check result suggesting persistent object caching.
+	 */
+	public function get_test_health_persistent_object_cache() {
+		/**
+		 * Filter the action URL for the persistent object cache health check.
+		 *
+		 * @since 6.1
```
Comment (audrasjb) [inline src/wp-admin/includes/class-wp-site-health.php]: `@since 6.1.0` would be better

File: src/wp-admin/includes/class-wp-site-health.php [hunk 4]
```diff
@@ -2253,6 +2253,104 @@ public function get_test_authorization_header() {
...
+		if ( ! empty( $available_services ) ) {
+			$notes .= ' ' . sprintf(
+				/* translators: Available object caching services. */
+				__( 'Your host appears to support the following object caching services: %s.' ),
+				implode( ', ', $available_services )
+			);
+		}
+
+		/**
+		 * Filter the second paragraph of the health check's description
```
Comment (audrasjb) [inline src/wp-admin/includes/class-wp-site-health.php]: "Filters" :)

File: src/wp-admin/includes/class-wp-site-health.php [hunk 5]
```diff
@@ -2253,6 +2253,104 @@ public function get_test_authorization_header() {
...
+		 * when suggesting the use of a persistent object cache.
+		 *
+		 * Hosts may want to replace the notes to recommend their preferred object caching solution.
+		 *
+		 * Plugin authors may want to append notes (not replace) on why object caching is recommended for their plugin.
+		 *
+		 * @since 6.1
```
Comment (audrasjb) [inline src/wp-admin/includes/class-wp-site-health.php]: `@since 6.1.0` :)

File: src/wp-admin/includes/class-wp-site-health.php [hunk 6]
```diff
@@ -2848,4 +2953,128 @@ public function is_development_environment() {
 		return in_array( wp_get_environment_type(), array( 'development', 'local' ), true );
 	}
 
+	/**
+	 * Determines whether to suggest using a persistent object cache.
+	 *
+	 * @since 6.1
```
Comment (audrasjb) [inline src/wp-admin/includes/class-wp-site-health.php]: `@since 6.1.0` :)

**Review by audrasjb — CHANGES_REQUESTED** [review]
I reported a few nitpicks on the PR (I didn't mention all occurrences though) 😇

**Comment by manuelRod** [comment]
Thanks for the nitpicks @audrasjb, I've fixed them.

**Review by audrasjb — APPROVED** [review]

**Review by spacedmonkey — APPROVED** [review]
Looks good to me.

**Comment by manuelRod** [comment]
I think this one is ready to be merged.
cc: @felixarntz

File: src/wp-admin/includes/class-wp-site-health.php [hunk 7]
```diff
@@ -2253,6 +2253,104 @@ public function get_test_authorization_header() {
...
+		 * Hosts may want to replace the notes to recommend their preferred object caching solution.
+		 *
+		 * Plugin authors may want to append notes (not replace) on why object caching is recommended for their plugin.
+		 *
+		 * @since 6.1.0
+		 *
+		 * @param string $notes              The notes appended to the health check description.
+		 * @param array  $available_services The list of available persistent object cache services.
+		 */
+		$notes = apply_filters( 'site_status_persistent_object_cache_notes', $notes, $available_services );
```
Comment (Clorith) [inline src/wp-admin/includes/class-wp-site-health.php:2335]: This filter may not be needed, since all tests can have each individual field filtered either via PHP, or JavaScript, with the `site_status_test_result` filter (which is ran on each individual test).

Unless there's a specific need for a unique one for this test specifically? If that's the case, perhaps it may even be better to expand on the existing filter to also have one that includes the test name in it for more targeted filters, such as for example `site_status_{$test_name}_test_result` (in this case becoming `site_status_persistent_object_cache_test_result`) 🤔

File: src/wp-admin/includes/class-wp-site-health.php [hunk 8]
```diff
@@ -2253,6 +2253,104 @@ public function get_test_authorization_header() {
...
+
+		$result['status']         = 'recommended';
+		$result['label']          = __( 'You should use a persistent object cache' );
+		$result['badge']['color'] = 'orange';
```
Comment (Clorith) [inline src/wp-admin/includes/class-wp-site-health.php]: The badge should not have changing colors based on status, the color in core is used to indicate which area it belongs to. This is also a necessity for accessibility, where a color change alone is not sufficient, and the label would need to change with it if that were the case.

Instead, the `status` should change with the status of the test (which it does here, so this could be easily resolved by removing the badge color change).

File: src/wp-admin/includes/class-wp-site-health.php [hunk 9]
```diff
@@ -2373,6 +2471,13 @@ public static function get_tests() {
 			);
 		}
 
+		if ( 'production' === wp_get_environment_type() ) {
```
Comment (Clorith) [inline src/wp-admin/includes/class-wp-site-health.php:2476]: Could we add an inline comment here, just to explain why the tests are only done in certain environments (I know it makes sense to us now, but it doesn't hurt to have the _why_ of it documented for the future).

File: src/wp-admin/includes/class-wp-site-health.php [hunk 10]
```diff
@@ -2373,6 +2471,13 @@ public static function get_tests() {
...
+			$tests['direct']['persistent_object_cache'] = array(
```
Comment (Clorith) [inline src/wp-admin/includes/class-wp-site-health.php:2477]: Should this perhaps be an `async` test, given the potential DB load from `should_suggest_persistent_object_cache()`?

File: src/wp-admin/includes/class-wp-site-health.php [hunk 11]
```diff
@@ -2848,4 +2953,128 @@ public function is_development_environment() {
...
+		if ( $thresholds['alloptions_bytes'] < strlen( serialize( $alloptions ) ) ) {
+			return true;
+		}
+
+		$table_names = implode( "','", array( $wpdb->comments, $wpdb->options, $wpdb->posts, $wpdb->terms, $wpdb->users ) );
+
+		// With InnoDB the `TABLE_ROWS` are estimates, which are accurate enough and faster to retrieve than individual `COUNT()` queries.
+		$results = $wpdb->get_results(
+			$wpdb->prepare(
+			// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared
```
Comment (Clorith) [inline src/wp-admin/includes/class-wp-site-health.php]: I'd love for any line that disables coding standard checks to include a comment on _why_ the standards are being ignored, and why that is expected. I couldn't find any guidelines on this, and am checking in with the standards team to make sure we do this the right way :)

**Review by Clorith — CHANGES_REQUESTED** [review]
Had a look over, and I have some questions that I left in-line, and a quick reference for some potential improvements.

File: src/wp-admin/includes/class-wp-site-health.php (hunk 7 above)
Comment (manuelRod) [reply src/wp-admin/includes/class-wp-site-health.php:2335]: Hello, @Clorith Thanks for chiming in. Good point here. I'm gonna try to explain the reason why behind it. 
We are porting these modules from the [performance plugin](https://github.com/WordPress/performance/), on there, we decided to have as many filters as needed to allow customization to this level. As we were aiming for hosting especially to active use these filters to give targeted recommendations to their clients.

File: src/wp-admin/includes/class-wp-site-health.php (hunk 8 above)
Comment (manuelRod) [reply src/wp-admin/includes/class-wp-site-health.php]: I didn't know that, so status is enough to get the right color. ok!

File: src/wp-admin/includes/class-wp-site-health.php (hunk 10 above)
Comment (manuelRod) [reply src/wp-admin/includes/class-wp-site-health.php:2477]: bringing audrasjb here, since we wrote originally this module.

File: src/wp-admin/includes/class-wp-site-health.php (hunk 7 above)
Comment (tillkruss) [reply src/wp-admin/includes/class-wp-site-health.php:2335]: @Clorith: This filter specifically was requested by @spacedmonkey or @felixarntz IIRC to allow hosting companies to easily add their own notes to this heath check about their object caching solution.

File: src/wp-admin/includes/class-wp-site-health.php (hunk 10 above)
Comment (tillkruss) [reply src/wp-admin/includes/class-wp-site-health.php:2477]: @manuelRod: I wrote this 🙃

Async makes sense.

File: src/wp-admin/includes/class-wp-site-health.php (hunk 9 above)
Comment (tillkruss) [inline src/wp-admin/includes/class-wp-site-health.php:2476]: Makes sense.

>Only check for a persistent object cache in production environments to not unnecessarily promote complicated setups

File: src/wp-admin/includes/class-wp-site-health.php (hunk 8 above)
Comment (tillkruss) [reply src/wp-admin/includes/class-wp-site-health.php]: @manuelRod: Can you update the PR, or do you want me to open on in the perflab repo first?

File: src/wp-admin/includes/class-wp-site-health.php (hunk 10 above)
Comment (manuelRod) [reply src/wp-admin/includes/class-wp-site-health.php:2477]: ohh @tillkruss sorry for that, i mixed who wrote what :)

File: src/wp-admin/includes/class-wp-site-health.php (hunk 8 above)
Comment (manuelRod) [reply src/wp-admin/includes/class-wp-site-health.php:2337]: @tillkruss should we do this first on the plugin level and then on the core level? or should we iterate directly here?

File: src/wp-admin/includes/class-wp-site-health.php (hunk 10 above)
Comment (manuelRod) [reply src/wp-admin/includes/class-wp-site-health.php:2477]: @tillkruss should we do this kind of big change, first on the plugin level, and then move them here?

File: src/wp-admin/includes/class-wp-site-health.php (hunk 8 above)
Comment (manuelRod) [reply src/wp-admin/includes/class-wp-site-health.php]: @tillkruss @Clorith this is also being discussed in [here](https://github.com/WordPress/wordpress-develop/pull/2894#discussion_r941100596). We should get a decision that will affect both Site Health ( and future ) modules.

File: src/wp-admin/includes/class-wp-site-health.php (hunk 11 above)
Comment (manuelRod) [reply src/wp-admin/includes/class-wp-site-health.php]: I removed the second ignore which was not needed. However, this one is needed for 
`AND TABLE_NAME IN ('$table_names')`

@Clorith let me know when you hear from the standards team.

Comment (tillkruss) [reply src/wp-admin/includes/class-wp-site-health.php]: @manuelRod: You removed the re-enable comment, which needs to stay.

**Review by tillkruss — CHANGES_REQUESTED** [review]

File: src/wp-admin/includes/class-wp-site-health.php (hunk 10 above)
Comment (tillkruss) [reply src/wp-admin/includes/class-wp-site-health.php:2477]: Yeah, maybe we do the badge color and async change on `wordpress/performance` first? Do you want to open a PR, or shall I?

File: src/wp-admin/includes/class-wp-site-health.php (hunk 11 above)
Comment (Clorith) [reply src/wp-admin/includes/class-wp-site-health.php]: There's no hard guideline on requiring omments per now, but the team agreed that this is something that should be created, so let's move forward with adding a comment on why a rule is being ignored here, be the change we want to see in the world 😃 

As for enabling/disabling of rules, you could also do a single-line-allow by using `// phpcs:ignore <rule name> -- <comment with reason for ignoring the rule right here>` for the single `$wpdb->prepare()` command, that may also be handy to not accidentally forget to re-enable a rule later on.

File: src/wp-admin/includes/class-wp-site-health.php (hunk 7 above)
Comment (Clorith) [reply src/wp-admin/includes/class-wp-site-health.php:2335]: Makes sense in general, I'm just curious why only this test would need that special filter, is this something that should be generalized for all tests instead? And is there any reasoning for not using the existing filters (like what shortcomings were encountered, these may be shortcomings that exist for the other existing tests that we need to consider as well elsewhere).

File: src/wp-admin/includes/class-wp-site-health.php (hunk 11 above)
Comment (tillkruss) [reply src/wp-admin/includes/class-wp-site-health.php]: @Clorith The wpdb query is extremely fast even on massive sites, I don't think we need to make this an async method.

File: src/wp-admin/includes/class-wp-site-health.php (hunk 7 above)
Comment (tillkruss) [reply src/wp-admin/includes/class-wp-site-health.php:2335]: The `site_status_persistent_object_cache_notes` is specifically for hosting providers wanting to _only_ alter the health check `description` if the test is suggestion the use of an object cache. Using standard filters would make this quite verbose.

File: src/wp-admin/includes/class-wp-site-health.php (hunk 11 above)
Comment (tillkruss) [reply src/wp-admin/includes/class-wp-site-health.php]: @manuelRod: Let's add the reason then. This query cannot use interpolation.

File: src/wp-admin/includes/class-wp-site-health.php (hunk 10 above)
Comment (manuelRod) [reply src/wp-admin/includes/class-wp-site-health.php:2477]: @Clorith as @tillkruss comments in the next comment block, this query is super fast, I don't think we need to move to async either.

File: src/wp-admin/includes/class-wp-site-health.php (hunk 11 above)
Comment (manuelRod) [reply src/wp-admin/includes/class-wp-site-health.php]: Ignore comment added.

**Comment by manuelRod** [comment]
@tillkruss @Clorith I've implemented the changes.
1. Removed badge color for different statuses.
2. Added ignore comment for interpolation/CS.
3. Do we agree that the tests stay direct instead of async? the query is light and fast.
4. Do we agree on the extra filters to allow more flexibility to hosting companies?

Thanks!

File: src/wp-admin/includes/class-wp-site-health.php (hunk 8 above)
Comment (manuelRod) [reply src/wp-admin/includes/class-wp-site-health.php]: fixed

File: src/wp-admin/includes/class-wp-site-health.php (hunk 11 above)
Comment (manuelRod) [reply src/wp-admin/includes/class-wp-site-health.php]: fixed

File: src/wp-admin/includes/class-wp-site-health.php (hunk 7 above)
Comment (felixarntz) [reply src/wp-admin/includes/class-wp-site-health.php:2335]: +1 on keeping this filter to make it obvious and straightforward for e.g. hosting providers that would like to alter the notes. The general filter would technically work as you're outlining @Clorith, however I agree that this would be overly verbose. The notes in this Site Health check are explicitly intended to be filtered, similarly to how e.g. the PHP upgrade notice that was separately implemented a few years back allows its additional explanation to be filtered. When we eventually document the ability to filter the notes, it would be unnecessarily complex and confusing to have to explain how you have to look for the specific field to modify. In addition, by having this specific filter, we maintain the same security standard as for the default content, as the content will go through `wp_kses()` _after_ filtering. If developers used the general filter, they would have to take care of it yourself, essentially duplicating what core already does here - or worse (and maybe even more likely), they don't know and don't even use `wp_kses()`.

File: src/wp-admin/includes/class-wp-site-health.php (hunk 8 above)
Comment (felixarntz) [reply src/wp-admin/includes/class-wp-site-health.php]: @manuelRod @tillkruss @Clorith I've opened https://github.com/WordPress/performance/issues/471 and https://github.com/WordPress/performance/pull/472 to fix this in the Performance Lab plugin throughout all Site Health checks.

File: src/wp-admin/includes/class-wp-site-health.php (hunk 10 above)
Comment (felixarntz) [reply src/wp-admin/includes/class-wp-site-health.php:2477]: Agreed, while I'm not opposed to making the check async, I don't think it's worth it either. Looking at existing Site Health checks, it appears like the policy for async tests is mostly around any external HTTP requests, which makes sense. Cheap database lookups like here however don't justify the additional effort IMO.

File: src/wp-admin/includes/class-wp-site-health.php [hunk 12]
```diff
@@ -2373,6 +2470,14 @@ public static function get_tests() {
 			);
 		}
 
+		// Only check for a persistent object cache in production environments to not unnecessarily promote complicated setups.
+		if ( 'production' === wp_get_environment_type() ) {
+			$tests['direct']['persistent_object_cache'] = array(
+				'label' => __( 'Persistent object cache' ),
+				'test'  => 'health_persistent_object_cache',
```
Comment (felixarntz) [inline src/wp-admin/includes/class-wp-site-health.php]: The `health_` prefix here and elsewhere does not make sense, let's remove it throughout this name and the relevant methods.
```suggestion
				'test'  => 'persistent_object_cache',
```

File: src/wp-admin/includes/class-wp-site-health.php [hunk 13]
```diff
@@ -2253,6 +2253,103 @@ public function get_test_authorization_header() {
...
 	}
 
+	/**
+	 * Tests persistent_object_cache health check.
+	 *
+	 * @since 6.1.0
+	 *
+	 * @return array The health check result suggesting persistent object caching.
+	 */
+	public function get_test_health_persistent_object_cache() {
```
Comment (felixarntz) [inline src/wp-admin/includes/class-wp-site-health.php]: See my other comment:
```suggestion
	public function get_test_persistent_object_cache() {
```

File: src/wp-admin/includes/class-wp-site-health.php [hunk 14]
```diff
@@ -2253,6 +2253,103 @@ public function get_test_authorization_header() {
 		return $result;
 	}
 
+	/**
+	 * Tests persistent_object_cache health check.
```
Comment (felixarntz) [inline src/wp-admin/includes/class-wp-site-health.php]: This is not really an appropriate documentation :)

Please describe in one or two sentences what this really does, not just the name of the test it refers to, which is a bit pointless.

File: src/wp-admin/includes/class-wp-site-health.php [hunk 15]
```diff
@@ -2253,6 +2253,103 @@ public function get_test_authorization_header() {
...
+	 *
+	 * @since 6.1.0
+	 *
+	 * @return array The health check result suggesting persistent object caching.
```
Comment (felixarntz) [inline src/wp-admin/includes/class-wp-site-health.php]: It won't necessarily suggest persistent object caching - it might also say that everything is fine. I would suggest we rephrase it to the same description that is used in other core methods for Site Health tests.
```suggestion
	 * @return array The test result.
```

File: src/wp-admin/includes/class-wp-site-health.php [hunk 16]
```diff
@@ -2253,6 +2253,103 @@ public function get_test_authorization_header() {
...
+		 * @since 6.1.0
+		 *
+		 * @param string $notes              The notes appended to the health check description.
+		 * @param array  $available_services The list of available persistent object cache services.
+		 */
+		$notes = apply_filters( 'site_status_persistent_object_cache_notes', $notes, $available_services );
+
+		$result['status']         = 'recommended';
+		$result['label']          = __( 'You should use a persistent object cache' );
+		$result['description']   .= sprintf(
```
Comment (felixarntz) [inline src/wp-admin/includes/class-wp-site-health.php]: I believe these are incorrectly indented? There are some additional spaces that should be removed.

**Review by felixarntz — COMMENTED** [review]
@manuelRod Excellent work, this looks solid to me. I left a few comments on minor things to improve, such as the test slug and some documentation.

File: src/wp-admin/includes/class-wp-site-health.php [hunk 17]
```diff
@@ -2848,4 +2953,123 @@ public function is_development_environment() {
...
+	}
+
+	/**
+	 * Returns a list of available persistent object cache services.
+	 *
+	 * @since 6.1.0
+	 *
+	 * @return array The list of available persistent object cache services.
+	 */
+	function available_object_cache_services() {
```
Comment (felixarntz) [inline src/wp-admin/includes/class-wp-site-health.php]: Methods should preferably always have an access modifier. Let's make this one `private`.

**Comment by manuelRod** [comment]
I think this is good to go now! cc: @Clorith @tillkruss @felixarntz

**Comment by tillkruss** [comment]
LGTM

File: src/wp-admin/includes/class-wp-site-health.php [hunk 18]
```diff
@@ -2253,6 +2253,105 @@ public function get_test_authorization_header() {
...
+
+		if ( ! $this->should_suggest_persistent_object_cache() ) {
+			$result['label'] = __( 'A persistent object cache is not required' );
+
+			return $result;
+		}
+
+		$available_services = $this->available_object_cache_services();
+
+		$notes = __( 'Your hosting provider can tell you if persistent object caches are able to be enabled on your site.' );
```
Comment (felixarntz) [inline src/wp-admin/includes/class-wp-site-health.php]: This sounds a bit too complicated.
```suggestion
		$notes = __( 'Your hosting provider can tell you if persistent object caches can be enabled on your site.' );
```

**Review by felixarntz — APPROVED** [review]
@manuelRod One tiny nit-pick, however I can address that one pre-commit. This looks good to go, great work!

**Comment by felixarntz** [comment]
@Clorith Can you please give this another review this week or early next week? It now has 3 approvals, from my perspective it's ready to commit.

File: src/wp-admin/includes/class-wp-site-health.php (hunk 18 above)
Comment (tillkruss) [reply src/wp-admin/includes/class-wp-site-health.php]: @felixarntz: Either `persistent object caching` or `a persistent object cache`.

Comment (manuelRod) [reply src/wp-admin/includes/class-wp-site-health.php]: "Your hosting provider can tell you if persistent object cache can be enabled on your site."

File: src/wp-admin/includes/class-wp-site-health.php [hunk 19]
```diff
@@ -2253,6 +2253,105 @@ public function get_test_authorization_header() {
...
+
+		if ( ! $this->should_suggest_persistent_object_cache() ) {
+			$result['label'] = __( 'A persistent object cache is not required' );
+
+			return $result;
+		}
+
+		$available_services = $this->available_object_cache_services();
+
+		$notes = __( 'Your hosting provider can tell you if persistent object cache can be enabled on your site.' );
```
Comment (tillkruss) [inline src/wp-admin/includes/class-wp-site-health.php:2312]: To me this sounds incorrect.

```suggestion
		$notes = __( 'Your hosting provider can tell you if a persistent object cache can be enabled on your site.' );
```

**Review by tillkruss — CHANGES_REQUESTED** [review]

**Review by Clorith — APPROVED** [review]
I agree with the small string change by @tillkruss, but other than that it looks good 👍

File: src/wp-admin/includes/class-wp-site-health.php [hunk 20]
```diff
@@ -2848,4 +2955,123 @@ public function is_development_environment() {
...
+		}
+
+		/**
+		 * Filters to force suggestion to use a persistent object cache and bypass threshold checks.
+		 *
+		 * @since 6.1.0
+		 *
+		 * @param bool $suggest Whether to suggest using a persistent object cache.
+		 */
+		if ( apply_filters( 'site_status_suggest_persistent_object_cache', false ) ) {
```
Comment (felixarntz) [inline src/wp-admin/includes/class-wp-site-health.php:2981]: @manuelRod A bit late, but I am just realizing that this should rather follow a more appropriate short-circuiting behavior: Right now, the only way to short-circuit would be to actually recommend a persistent object cache. But it also needs to be possible to run custom logic that would result in _not_ recommending a persistent object cache. So I think the default value here should be `null`, and any boolean result should cause the method to return early with that result.

This is a trivial change, which is compatible with the current behavior, it just expands it to also allow returning `false` to short-circuit.

**Comment by felixarntz** [comment]
@manuelRod Committed in https://core.trac.wordpress.org/changeset/53955 🎉
